Models, Files and Fine-Tuning APIs

thehappydinoa opened the pull request:
Refactored each API into it's own file and added three new APIs. I also removed the `Engines` API as it is now deprecated.

**Changes**
- [x] Implement Models API
- [x] Implement Files API
- [x] Implement Fine-Tune API
- [x] Write tests for the Models API
- [x] Write tests for the Files API
- [x] Write tests for the Fine-Tune API

Commit:
feat: add the files and fine tune apis

Commit:
chore: various cleanup and linting

Commit:
chore: add gitignore

Commit:
fix: null pointer exception

Commit:
chore: refactor

Commit:
chore: boilerplate tests

Commit:
chore: update tests

thehappydinoa commented on client.go:
```diff
@@ -0,0 +1,139 @@
+	}
+
+	for _, option := range options {
+		if err := option(c); err != nil {
```
I am super open to suggestions, I didn't want to modify too much.

thehappydinoa commented on client.go:
```diff
@@ -0,0 +1,139 @@
+	return c, nil
+}
+
+func (c *client) newRequest(ctx context.Context, method, path string, payload interface{}) (*http.Request, error) {
```
This was copied verbatim from gpt3.go, maybe items like that would be better for another MR.

thehappydinoa commented on client.go:
```diff
@@ -0,0 +1,139 @@
+	return resp, checkForSuccess(resp)
+}
+
+func checkForSuccess(resp *http.Response) error {
```
Unfortunately the stream and file contents methods do not return json. I left it broken out so only methods that need it can use it.

thehappydinoa commented on edits.go:
```diff
@@ -0,0 +1,57 @@
+	TotalTokens      int `json:"total_tokens"`
+}
+
+func (c *client) Edits(ctx context.Context, request EditsRequest) (*EditsResponse, error) {
```
Honestly my goal was to keep as much backwards compatibility as possible. Maybe we can address renaming existing methods in another MR.

tylermann replied:
yeah, I think it is probably fine as it is and already being used publicly. we could alias it if we want in another PR, but I'd prefer not to rename.

thehappydinoa commented on files.go:
```diff
@@ -0,0 +1,135 @@
+// Files lists the files that belong to the user's organization.
+//
+// See: https://beta.openai.com/docs/api-reference/files/list
+func (c *client) Files(ctx context.Context) (*FilesResponse, error) {
```
Oh I would like it if they all had names like that, but I saw Engines and Completions and didn't want to break the pattern.

tylermann replied:
yeah I think this is probably better to just match the existing naming convention. although I don't feel that strongly either way for new methods.

thehappydinoa commented:
Understood. Just for context I didn't actually create the name scheme, I just copied in the one from gpt3.go. The same goes for the base client functionality. I am open to changing some of those, but I wanted to clarify if this MR is the right place. My main goal was to enable folks to use the new completion and fine tune endpoints. Please let me know how we should proceed, I would love to be able to use some of these features.

tylermann commented on completions.go:
```diff
@@ -0,0 +1,146 @@
+
+func (c *client) CompletionStream(ctx context.Context, request CompletionRequest, onData func(*CompletionResponse)) error {
+	request.Stream = true
+	req, err := c.newRequest(ctx, "POST", "/completions", request)
```
we probably want to apply the default model here the same as we do for `Completion` above.

tylermann commented:
> Understood. Just for context I didn't actually create the name scheme, I just copied in the one from gpt3.go. The same goes for the base client functionality. I am open to changing some of those, but I wanted to clarify if this MR is the right place. My main goal was to enable folks to use the new completion and fine tune endpoints. Please let me know how we should proceed, I would love to be able to use some of these features.
> 
> Also maybe @tylermann can weigh in here?

Yeah I would prefer as minimal changes as possible to the existing APIs since this is a publicly used/open source library and will break when people upgrade. Andy (pullrequest), would you mind taking another pass at this review with the mindset of preserving existing functionality? I believe the main goal here is to fulfill this issue: https://github.com/PullRequestInc/go-gpt3/issues/12

tylermann commented on embeddings.go:
```diff
@@ -0,0 +1,84 @@
+//
+// See: https://beta.openai.com/docs/api-reference/embeddings
+func (c *client) Embeddings(ctx context.Context, request EmbeddingsRequest) (*EmbeddingsResponse, error) {
+	req, err := c.newRequest(ctx, "POST", "/embeddings", request)
```
yeah we can punt on this if needed, but could potentially do something like:
```
output := EmbeddingResponse{}
return c.post("/embeddings", request, &output)
```

tylermann commented on fine_tunes.go:
```diff
@@ -0,0 +1,279 @@
+		}
+		output := new(FineTuneEvent)
+		if err := json.Unmarshal(line, output); err != nil {
+			return fmt.Errorf("invalid json stream data: %v", err)
```
it could be nice to share the code of the streaming of fine tunes with completion streaming. just becuase I noticed some of the handling is slightly different and this error output is different. that being said maybe it isn't worth the effort for only two repetitions of the code so far.

thehappydinoa replied:
Yeah I completely agree and would love to have a shared set of methods for them. I am sure OpenAI is just going to keep adding more streaming APIs, hopefully the ChatGPT API?

tylermann commented on completions.go:
```diff
@@ -0,0 +1,146 @@
+
+		output := new(CompletionResponse)
+		if err := json.Unmarshal(line, output); err != nil {
+			return err
```
this used to use a `fmt.Errorf` like the version used in the fine tuning API. we may want to continue doing the previous behavior just in case anyone was expecting it, but also for consistency.

tylermann reviewed (commented):
@thehappydinoa , first off thank you for the work here to improve this library and update these APIs!

I am slightly worried about the size of this PR just because it does change a lot of lines at once even if many of them are just moved from other files. Perhaps we could split those things up next time if we want to reorder and do that in a separate PR so it easier to follow the actual changes.

I know there are some intentional breaking changes as a part of this PR as we are removing some deprecated methods. Although ideally we can try to preserve as much functionality as possible especially for the most likely used methods like the Completion ones.

I left some comments inline, but I'll also play around with this to help manually test it.

thehappydinoa commented:
Hey @tylermann thank you for taking the time to review my pr. Definitely something that should've split up. I will go your comments and try to implement what I can!

Thanks again!

thehappydinoa commented on client.go:
```diff
@@ -0,0 +1,139 @@
+var dataPrefix = []byte("data: ")
+var streamTerminationPrefix = []byte("[DONE]")
+
+type Client interface {
```
Thats totally fair, let me add those back in.

Commit:
fix: apply feedback

thehappydinoa commented:
Hey there, 
Thank you both for taking the time to review my changes. I really appreciate the feedback and I implemented everything except the situation when the body is `null` as delete and a couple others have no response body. I hope that I addressed all your concerns.

Thanks again,
Aidan

tylermann commented on client.go:
```diff
@@ -0,0 +1,143 @@
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
```
I was testing out the changes on this branch locally and was running into errors, that I believe are from this line. I don't think we used to do this defer/body close from the `performRequest` happy path, so it was giving me an error when trying to run `go run cmd/test/main.go`
> invalid json response: http2: response body closed

thehappydinoa replied:
Ah let me fix that, I believe I was told to move it there, does this occur when there is no response body?

tylermann reviewed (changes requested):
thanks for the updates here @thehappydinoa , I was testing this branch locally and ran into a few errors. I will keep playing around with it but I think this current issue is preventing even the happy path for Completion requests from working for me.

Jekyle9444 reviewed (approved):
[Můj_QR_1720745704212.csv](https://github.com/user-attachments/files/16186435/Muj_QR_1720745704212.csv)
